### PR TITLE
Phrase grounding datasets/modules/transforms

### PR DESCRIPTION
**Pull request by ebsmothers**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #200
* #199
* #198
* #181
* #180
* __->__ #179

Differential Revision: [D38160121](https://our.internmc.facebook.com/intern/diff/D38160121)

### CODE REVIEW

File: examples/mdetr/data/modulated_detection.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+from examples.mdetr.data.transforms import ConvertCocoPolysToMask, create_positive_map
+from torchvision.datasets import CocoDetection
+
+
+class ModulatedDetection(CocoDetection):
+    """The base dataset class for most MDETR datasets."""
```
Comment (RdoubleA) [inline examples/mdetr/data/modulated_detection.py:12]: some quick details on what this adds?

File: examples/mdetr/data/transforms.py [hunk 2]
```diff
@@ -0,0 +1,459 @@
...
+from torchvision.ops.boxes import box_convert
+
+
+def crop(image, target, region):
+    cropped_image = F.crop(image, *region)
+
+    target = target.copy()
+    i, j, h, w = region
+
+    # should we do something wrt the original size?
```
Comment (RdoubleA) [inline examples/mdetr/data/transforms.py:25]: is this a comment from the original codebase? should we remove this?

File: examples/mdetr/data/transforms.py [hunk 3]
```diff
@@ -0,0 +1,459 @@
...
+        cropped_boxes = boxes - torch.as_tensor([j, i, j, i])
+        cropped_boxes = torch.min(cropped_boxes.reshape(-1, 2, 2), max_size)
+        cropped_boxes = cropped_boxes.clamp(min=0)
+        area = (cropped_boxes[:, 1, :] - cropped_boxes[:, 0, :]).prod(dim=1)
+        target["boxes"] = cropped_boxes.reshape(-1, 4)
+        target["area"] = area
+        fields.append("boxes")
+
+    if "masks" in target:
+        # FIXME should we update the area here if there are no boxes?
```
Comment (RdoubleA) [inline examples/mdetr/data/transforms.py:42]: should this FIXME stay here?

File: examples/mdetr/data/transforms.py [hunk 4]
```diff
@@ -0,0 +1,459 @@
...
+        self, image: Union[Iterable[Image], Image], target: Dict[str, Any]
+    ) -> torch.Tensor:
+        image, target = self.image_transform(image, target)
+        target["tokenized"] = self.tokenizer(target["caption"], return_tensors="pt")[
+            "input_ids"
+        ][0]
+        return image, target
+
+
+def create_positive_map(tokenized, tokens_positive):
```
Comment (RdoubleA) [inline examples/mdetr/data/transforms.py:342]: can you just use the `construct_positive_map` from your contrastive alignment loss?

File: examples/mdetr/utils/misc.py [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+        cur_count = 0
+        for v in batch[1]:
+            cur_pos = v["positive_map"]
+            batched_pos_map[
+                cur_count : cur_count + len(cur_pos), : cur_pos.shape[1]
+            ] = cur_pos
+            cur_count += len(cur_pos)
+
+        assert cur_count == len(batched_pos_map)
+        # assert batched_pos_map.sum().item() == sum([v["positive_map"].sum().item() for v in batch[1]])
```
Comment (RdoubleA) [inline examples/mdetr/utils/misc.py:34]: remove this line?

File: examples/mdetr/utils/misc.py [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+        cur_count = 0
+        for v in batch[1]:
+            cur_pos = v["positive_map_eval"]
+            batched_pos_map[
+                cur_count : cur_count + len(cur_pos), : cur_pos.shape[1]
+            ] = cur_pos
+            cur_count += len(cur_pos)
+
+        assert cur_count == len(batched_pos_map)
+        # assert batched_pos_map.sum().item() == sum([v["positive_map"].sum().item() for v in batch[1]])
```
Comment (RdoubleA) [inline examples/mdetr/utils/misc.py:52]: remove this line?

**Review by RdoubleA — APPROVED** [review]

File: examples/mdetr/data/transforms.py (hunk 3 above)
Comment (ebsmothers) [reply examples/mdetr/data/transforms.py:42]: We can remove it. I may also remove the mask logic in a separate PR but need to double-confirm that it is not used anywhere else.

File: examples/mdetr/data/transforms.py (hunk 4 above)
Comment (ebsmothers) [reply examples/mdetr/data/transforms.py:342]: Not as things are currently written. That one relies on the indices output by the matcher to order the positive map for the loss calculation. I have a to-do to try and refactor to a common implementation, but it is low-pri at the moment relative to other things.

**Comment by ebsmothers** [comment]
@ebsmothers has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38160121).

**Comment by ebsmothers** [comment]
@ebsmothers has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38160121).

**Comment by ebsmothers** [comment]
@ebsmothers has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38160121).

File: examples/mdetr/data/datamodule.py [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+
+
+class FlickrDataModule(LightningDataModule):
+    def __init__(self, dataset_config, tokenizer: Optional[Callable] = None):
+        super().__init__()
+        self.dataset_config = dataset_config
+        self.distributed = dataset_config.distributed
+        self.batch_size = dataset_config.batch_size
+        self.tokenizer = tokenizer
+        self.num_workers = 0
```
Comment (ankitade) [inline examples/mdetr/data/datamodule.py:26]: remove and fall back to default

File: examples/mdetr/data/transforms.py [hunk 8]
```diff
@@ -0,0 +1,457 @@
...
+
+    def __init__(self, tokenizer: Callable, is_train: bool, cautious: bool = True):
+        normalize = Compose(
+            [ToTensor(), Normalize(self.IMAGENET_MEAN, self.IMAGENET_STD)]
+        )
+        self.tokenizer = tokenizer
+        scales = [480, 512, 544, 576, 608, 640, 672, 704, 736, 768, 800]
+        max_size = 1333
+        if is_train:
+            horizontal = [] if cautious else [RandomHorizontalFlip()]
```
Comment (ankitade) [inline examples/mdetr/data/transforms.py:304]: what does cautious mean?

File: examples/mdetr/utils/misc.py [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+            input.transpose(0, 1), size, scale_factor, mode, align_corners
+        ).transpose(0, 1)
+
+    # empty batch dimension is now supported in pytorch
+    return torch.nn.functional.interpolate(
+        input, size, scale_factor, mode, align_corners
+    )
+
+
+def targets_to(targets: List[Dict[str, Any]], device):
```
Comment (ankitade) [inline examples/mdetr/utils/misc.py:106]: used where?

File: examples/mdetr/data/flickr.py [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+from pathlib import Path
+
+from examples.mdetr.data.modulated_detection import ModulatedDetection
+
+
+def build_flickr(image_set, tokenizer, transform, args):
```
Comment (ankitade) [inline examples/mdetr/data/flickr.py:12]: just make one dataset.py and move this, modulatedDataset and  collate fn there?

File: examples/mdetr/utils/misc.py (hunk 9 above)
Comment (ebsmothers) [reply examples/mdetr/utils/misc.py:106]: Used later in the stack (e.g. phrase_grounding.py in #181 )

File: examples/mdetr/data/transforms.py (hunk 8 above)
Comment (ebsmothers) [reply examples/mdetr/data/transforms.py:304]: It's just referring to the degree of randomness in the transform. Actually it's only used for CLEVR, since we're not supporting that I will remove it

**Comment by ebsmothers** [comment]
@ebsmothers has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38160121).
